pkg/abstract/map.go

```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package abstract
+
+// Map abstracts reading from a map from Key to Val.
+// The Key values must be comparable in some way
+// (which is not exposed in this interface).
+// The map may or may not vary over time.
+// Access through this interface may or may not be thread-safe.
+type Map[Key, Val any] interface {
```

Comment:
This makes util.ConcurrentMap redundant.

Reply:
Do we have an independent need (aside for replacing ConcurrentMap) for this interface? Or is its sole purpose to serve MapLocker?

Reply:
This interface only provides read access, not write access. I have found it useful in the past (e.g., in the legacy system - https://github.com/kubestellar/kubestellar/tree/v0.14.2/pkg/placement) to separate interfaces for reading and writing. Further, I have also found it useful to define interfaces for fragments of reading functionality.

The current (in this PR) three-layered implementation of `(*bindingPolicyResolution).GetWorkload` uses three layers that I have found to be useful independently and in different combinations in the past. `util.ConcurrentMap` is partly redundant with the combination of two of them. Defining the `Map` interface makes it possible to have various layers and combine them as needed for a given use.